### PR TITLE
Tamanna/linear interpolation

### PR DESCRIPTION
**Pull request by tamannak8**
Linear interpolation of keyframes

### CODE REVIEW

File: reconstruction/linear_interpolator/linear_interpolator.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+from utilities import ConfigHandler
+from utilities import CloudFunctionality
+import time
+
+
+def linear_interpolation(frame_a, frame_b, alpha):
+    return cv2.addWeighted(frame_a, 1 - alpha, frame_b, alpha, 0)
+
+
+def reconstruct_video_with_keyframe_images(target_frame_rate=30):
```
Comment (vinayakSharm) [inline reconstruction/linear_interpolator/linear_interpolator.py:23]: docstrings pls

File: reconstruction/linear_interpolator/linear_interpolator.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+).stdout.rstrip("\n")
+# add git repo path to use all libraries
+sys.path.append(root_path)
+
+from utilities import ConfigHandler
+from utilities import CloudFunctionality
+import time
+
+
+def linear_interpolation(frame_a, frame_b, alpha):
```
Comment (vinayakSharm) [inline reconstruction/linear_interpolator/linear_interpolator.py:19]: docstring

File: reconstruction/linear_interpolator/linear_interpolator.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
+import cv2
```
Comment (vinayakSharm) [inline reconstruction/linear_interpolator/linear_interpolator.py:1]: module docstring

File: reconstruction/linear_interpolator/linear_interpolator.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        video_name = Path(str(video)).stem
+
+        # Get video properties
+        frame_rate = int(cap.get(cv2.CAP_PROP_FPS))
+        frame_size = (
+            int(cap.get(cv2.CAP_PROP_FRAME_WIDTH)),
+            int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT)),
+        )
+        input_frame_count = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
+        num_interpolated_frames = int((target_frame_rate / frame_rate) - 1)
```
Comment (vinayakSharm) [inline reconstruction/linear_interpolator/linear_interpolator.py:37]: can you rename this variable its misleading. Its not the number of interpolated frames. Its actually the the number of interpolated frames per second.

Comment (vinayakSharm) [reply reconstruction/linear_interpolator/linear_interpolator.py:37]: more over, its a factor. So this difference og_fps, and the requested_fps allows you to determine the factor multiple by which to create frames

**Review by vinayakSharm — CHANGES_REQUESTED** [review]
hi tamanna small docstring request. 

also could you help me understand the algorithm irl :)

File: reconstruction/linear_interpolator/linear_interpolator.py (hunk 4 above)
Comment (tamannak8) [reply reconstruction/linear_interpolator/linear_interpolator.py]: updated

File: reconstruction/linear_interpolator/linear_interpolator.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    """
+    Reconstruct a video with keyframe images.
+
+    This function takes a set of video files from a directory of reduced videos and reconstructs
+    them by interpolating frames to achieve a target frame rate. It saves the reconstructed video
+    in the 'reconstructed_videos' directory.
+
+    Parameters:
+    ------------
+    - target_frame_rate: The desired frame rate for the reconstructed video (default is 30).
```
Comment (vinayakSharm) [inline reconstruction/linear_interpolator/linear_interpolator.py:51]: hi tamanna no "-" to create consistency in code base.

File: reconstruction/linear_interpolator/linear_interpolator.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+            ret, frame_b = cap.read()
+
+            if not ret:
+                break
+
+            # Perform linear interpolation
+            for i in range(interpolation_factor + 1):
+                alpha = i / (interpolation_factor + 1)
+                interpolated_frame = linear_interpolation(frame_a, frame_b, alpha)
+                out.write(interpolated_frame)
```
Comment (vinayakSharm) [inline reconstruction/linear_interpolator/linear_interpolator.py:89]: hi tamanna in the while loop my worry is that we are not writing frame_a and frame_b. Only the interpolated frame. Shouldn't we be writing all 3?

**Review by vinayakSharm — CHANGES_REQUESTED** [review]
Small change on the "-" in the docstring. 

Comment/concern on writing out only the interpolated frames and not frames a,b in for loop. (google chatted as well :) )

File: reconstruction/linear_interpolator/linear_interpolator.py (hunk 6 above)
Comment (tamannak8) [reply reconstruction/linear_interpolator/linear_interpolator.py:89]: if alpha is 0 then it automatically writes frame a and when alpha =1 it writes frame b
so it gets handled
if you see add weighted formula it alpha*frame b and (1- alpha)*framea

File: config.ini [hunk 7]
```diff
@@ -246,6 +262,13 @@ dynamodb_table = leto_yolo
 invocation_type = RequestResponse ; Use 'Event' for asynchronous invocation
 =======
 invocation_type = RequestResponse ; Use 'Event' for asynchronous invocation
+=======
```
Comment (vinayakSharm) [inline config.ini:265]: hi tamanna it seems as though from merging main into your branch there is still an unresolved merge conflict:

"<<<<<<< HEAD
invocation_type = RequestResponse ; Use 'Event' for asynchronous invocation
=======
invocation_type = RequestResponse ; Use 'Event' for asynchronous invocation
=======
invocation_type = RequestResponse ; Use 'Event' for asynchronous invocation
"

please clear this out :)

File: reconstruction/linear_interpolator/linear_interpolator.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+    Perform linear interpolation between two image frames.
+
+    This function blends two input frames together using linear interpolation.
+    The interpolation is controlled by the parameter 'alpha', where 0 <= alpha <= 1.
+    When alpha is 0, the result will be identical to 'frame_a', and when alpha is 1,
+    the result will be identical to 'frame_b'.
+
+    Parameters:
+     -----------
+    frame_a: The first input image frame (numpy.ndarray).
```
Comment (vinayakSharm) [inline reconstruction/linear_interpolator/linear_interpolator.py:30]: the docstrings are really close. 

they need an indent, and the type list afer the colon, with the description underneath

ex/ from superres
      
          Parameters
          ----------
              model_prefix_s3: string
                  the prefix of the pretrained model as stored in s3.
      
          Returns
          ----------
              split_on_scaling: string
                  defines which model to use for cv2.dnn_superres as parsed from the model_prefix_s3

could you please fix them for this python file.

**Review by vinayakSharm — CHANGES_REQUESTED** [review]
hi tamanna the logic for the linear interpolation has been cleared. Thanks for helping me out on friday to understand this. 

there are 2 things. I just caught a bug in your config.ini 

and the docstrings have been formatted with the "----" however they still need to be modified to achieve consistency with the rest of the code base

**Review by vinayakSharm — APPROVED** [review]
thanks!

**Pull request merged**